README.md
# uptimerobot-operator

This operator automatically creates uptime monitors at [UptimeRobot](https://uptimerobot.com) for your Kubernetes Ingress resources. This allows you to easily integrate uptime monitoring of your services into your Kubernetes deployments.

> :warning: **This project is still in an early phase. Use it on your own risk but make sure to create issues for issues you encounter.**

:heart: to UptimeRobot for providing a pro account to be able to test the pro plan features!

## Usage

After installing the uptimerobot-operator it'll watch for Ingress resources in your Kubernetes cluster. For example after creating the following Ingress

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: my-ingress
  annotations:
    uroperator.brennerm.github.io/monitor.type: HTTPS
    uroperator.brennerm.github.io/monitor.interval: "600"
spec:
  rules:
  - host: brennerm.github.io
    http:
      paths:
      - pathType: Prefix
        path: "/"
        backend:
          service:
            name: service1
            port:
              number: 80
```

a new monitor for the URL *https://brennerm.github.io* is automatically being created in your UptimeRobot account.

The same monitor can also be created using the UptimeRobotMonitor resource like so:

```yaml
apiVersion: uroperator.brennerm.github.io/v1beta1
kind: UptimeRobotMonitor
metadata:
  name: my-custom-monitor
spec:
  url: "https://brennerm.github.io"
  type: HTTPS
  interval: 600
```

The operator also supports creating public status pages. See below for details.

## Installation

### Create an UptimeRobot API key

1. [Create an](https://uptimerobot.com/signUp) or login to your UptimeRobot account (no credit card required and they provide up to 50 monitors for free)
2. Go to "My Settings"
3. Generate and save "Main API Key" (the other API keys do not provide sufficient permissions to create, update and delete monitors)

### Deploying to Kubernetes using Helm

1. Add the uptimerobot-operator chart repo `helm repo add uptimerobot-operator https://brennerm.github.io/uptimerobot-operator/helm`
2. Deploy the Helm chart `helm upgrade --install uptimerobot-operator uptimerobot-operator --set uptimeRobotApiKey=$MY_UPTIMEROBOT_API_KEY`

Have a look at the [values file](helm/uptimerobot-operator/values.yaml) if you want to customize the deployment.

### Running local

> :information_source: **The following commands will make the operator work with your currently selected Kubernetes cluster (`kubectl config current-context`).**

1. Install all dependencies `pipenv install`
2. Set UptimeRobot API key `export UPTIMEROBOT_API_KEY=$MY_UPTIMEROBOT_API_KEY`
3. Start operator `kopf run --standalone ur_operator/handlers.py`

### Running in self-built Docker

1. Build Docker image `docker build -t uptimerobot-operator .`
2. Start container `docker run -e UPTIMEROBOT_API_KEY=$MY_UPTIMEROBOT_API_KEY -v ~/.kube:/home/ur_operator/.kube uptimerobot-operator`

### Running in pre-built Docker

1. Start container `docker run -e UPTIMEROBOT_API_KEY=$MY_UPTIMEROBOT_API_KEY -v ~/.kube:/home/ur_operator/.kube ghcr.io/brennerm/uptimerobot-operator:latest`

## Documentation

### UptimeRobotMonitor

The UptimeRobotMonitor resource supports all current parameters for monitors that UptimeRobot offers. Below you can find a list that contains all of them.

|key|type|description|
|-|-|-|
|`url` (required)|`string`|URL that will be monitored|
|`type` (required)|`string`|Type of monitor, one of: HTTP,HTTPS,KEYWORD,PING,PORT,HEARTBEAT|
|`friendlyName`|`string`|Friendly name of monitor, defaults to name of UptimeRobotMonitor object|
|`subType`|`string`|Subtype of monitor, one of: HTTP,HTTPS,KEYWORD,PING,PORT,HEARTBEAT|
|`port`|`integer`|Port to monitor when using monitor sub type PORT|
|`keywordType`|`string`|Keyword type when using monitor type KEYWORD, one of: EXISTS,NOT_EXISTS|
|`keywordValue`|`string`|Keyword value when using monitor type KEYWORD|
|`interval`|`integer`|The interval for the monitoring check (300 seconds by default)|
|`httpUsername`|`string`|Used for password protected pages when using monitor type HTTP,HTTP or KEYWORD, deprecated: use httpAuthSecret|
|`httpPassword`|`string`|Used for password protected pages when using monitor type HTTP,HTTP or KEYWORD, deprecated: use httpAuthSecret|
|`httpAuthSecret`|`string`|reference to a Kubernetes secret in the same namespace containing user and password for password protected pages when using monitor type HTTP,HTTPS or KEYWORD|
|`httpAuthType`|`string`|Used for password protected pages when using monitor type HTTP,HTTPS or KEYWORD, one of: BASIC_AUTH,DIGEST|
|`httpMethod`|`string`|The HTTP method to be used, one of: HEAD,GET,POST,PUT,PATCH,DELETE,OPTIONS|
|`postType`|`string`|The format of data to be sent with POST, PUT, PATCH, DELETE, OPTIONS requests|
|`postContentType`|`string`|The Content-Type header to be sent with POST, PUT, PATCH, DELETE, OPTIONS requests, one of: TEXT_HTML,APPLICATION_JSON|
|`postValue`|`object`|The data to be sent with POST, PUT, PATCH, DELETE, OPTIONS requests|
|`customHttpHeaders`|`object`|Custom HTTP headers to be sent along monitor request, formatted as JSON|
|`customHttpStatuses`|`string`|Allows to define HTTP status codes that will be handled as up or down, e.g. 404:0_200:1 to accept 404 as down and 200 as up|
|`ignoreSslErrors`|`boolean`|Flag to ignore SSL certificate related issues|
|`alertContacts`|`string`|Alert contacts to be notified when monitor goes up or down. For syntax check https://uptimerobot.com/api/#newMonitorWrap|
|`mwindows`|`string`|Maintenance window IDs for this monitor|

### Ingress

For Ingress resources the same parameters are supported. You pass them through annotations attached to your Ingress with the prefix `uroperator.brennerm.github.io/monitor.`.
See below for an example.

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: my-ingress
  annotations:
    uroperator.brennerm.github.io/monitor.type: HTTPS
    uroperator.brennerm.github.io/monitor.interval: "600"
spec:
  rules:
...
```

To disable ingress handling completely pass the environment variable `URO_DISABLE_INGRESS_HANDLING=1` to the operator.

### Public Status Pages

The PublicStatusPage resource supports all current parameters for status pages that UptimeRobot offers. Below you can find a list that contains all of them.

|key|type|description|
|-|-|-|
|`monitors` (required)|`string`|the list of monitor IDs to be displayed in status page (the values are seperated with "-" or 0 for all monitors)|
|`friendlyName`|`string`|Friendly name of public status page, defaults to name of PublicStatusPage object|
|`customDomain`|`string`|the domain or subdomain that the status page will run on|
|`password`|`string`|the password for the status page, deprecated: use passwordSecret|
|`passwordSecret`|`string`|reference to a Kubernetes secret in the same namespace containing the password for the status page|
|`sort`|`string`|the sorting of the monitors on the status page, one of: FRIENDLY_NAME_A_Z,FRIENDLY_NAME_Z_A,STATUS_UP_DOWN_PAUSED,STATUS_DOWN_UP_PAUSED|
|`status`|`string`|the status of the status page, one of: PAUSED,ACTIVE|
|`hideUrlLinks`|`boolean`|Flag to remove the UptimeRobot link from the status page (pro plan feature)|

```yaml
apiVersion: uroperator.brennerm.github.io/v1beta1
kind: PublicStatusPage
metadata:
  name: my-public-status-page
spec:
  monitors: "0" # will include all monitors
```

### Maintenance Windows

The MaintenanceWindow resource supports all current parameters for maintenance windows that UptimeRobot offers. Below you can find a list that contains all of them.

|key|type|description|
|-|-|-|
|`type` (required)|`string`|the type of maintenance window, one of: ONCE,DAILY,WEEKLY,MONTHLY|
|`startTime` (required)|`string`|the start time of the maintenance window, in seconds since epoch for type MaintenanceWindowType.ONCE, in HH:mm format for the other types|
|`duration` (required)|`number`|the number of seconds the maintenance window will be active|
|`friendlyName`|`string`|friendly name of the maintenance window, defaults to name of the MaintenanceWindow object|
|`value`|`string`|allows to specify the maintenance window selection, e.g. 2-4-5 for Tuesday-Thursday-Friday or 10-17-26 for the days of the month, only valid and required for MaintenanceWindowType.WEEKLY and MaintenanceWindowType.MONTHLY|

```yaml
apiVersion: uroperator.brennerm.github.io/v1beta1
kind: MaintenanceWindow
metadata:
  name: my-maintenance-window
spec:
  type: DAILY
  startTime: "10:00"
  duration: 30
```

### Alert Contacts

The AlertContact resource supports all current parameters for alert contacts that UptimeRobot offers. Below you can find a list that contains all of them.

|key|type|description|
|-|-|-|
|`type` (required)|`string`|the type of alert contact, one of: SMS,EMAIL,TWITTER_DM,BOXCAR,WEB_HOOK,PUSHBULLET,ZAPIER,PUSHOVER,HIPCHAT,SLACK|
|`value` (required)|`string`|the alert contact's mail address / phone number / URL / connection string|
|`friendlyName`|`string`|friendly name of the alert contact, defaults to name of the AlertContact object|

```yaml
apiVersion: uroperator.brennerm.github.io/v1beta1
kind: AlertContact
metadata:
  name: my-alert-contact
spec:
  type: EMAIL
  value: foo@bar.com
```

## Planned features

- provide a Helm chart to ease deployment :heavy_check_mark:
- support all configuration parameters for Monitors that UptimeRobot offers :heavy_check_mark:
- add support for creating Uptime Robot :heavy_check_mark:
  - alert contacts, :heavy_check_mark:
  - maintenance windows :heavy_check_mark:
  - public status pages using Kubernetes resources :heavy_check_mark:
- implement automatic detection of HTTP path of Ingress resources
- add an integration for external-dns to support creating monitors for Service resources


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.4.0] - 2021-08-30

### Added

- new property `passwordSecret` to PublicStatusPage resource, allows to reference password from Kubernetes secret [#22](https://github.com/brennerm/uptimerobot-operator/pull/22)
- new property `httpAuthSecret` to UptimeRobotMonitor resource, allows to reference username and password from Kubernetes secret [#23](https://github.com/brennerm/uptimerobot-operator/pull/23)

### Deprecated

- `password` property of PublicStatusPage resource, use `passwordSecret` instead
- `httpUsername` and `httpPassword` property of UptimeRobotMonitor resource, use `httpAuthSecret` instead

## [v0.3.0] - 2021-02-16

### Added

- introducing the PublicStatusPage resource to setup UptimeRobot status pages [#15](https://github.com/brennerm/uptimerobot-operator/pull/15)
- introducing the MaintenanceWindow resource to setup UptimeRobot maintenance windows [#16](https://github.com/brennerm/uptimerobot-operator/pull/16)
- introducing the AlertContact resource to setup UptimeRobot alert contacts [#21](https://github.com/brennerm/uptimerobot-operator/pull/21)

## [v0.2.0] - 2021-02-08

### Added

- allow to disable creating/updating/deleting monitors for Ingress resources [#6](https://github.com/brennerm/uptimerobot-operator/pull/6)
- the Helm chart now enables and makes use of the operator liveness endpoint [#8](https://github.com/brennerm/uptimerobot-operator/pull/8)
- add support for all monitor paramters to the UptimeRobot monitor resource [#10](https://github.com/brennerm/uptimerobot-operator/pull/10)
- allow to set all monitor parameters through Ingress annotations [#13](https://github.com/brennerm/uptimerobot-operator/pull/13)

## [v0.1.0] - 2021-02-02

### Added

- initial functionality
  - create ping monitors for Ingress resources
  - allow to create custom monitor with UptimeRobotMonitor resource
- first version of Helm chart


ur_operator/crds/monitor.py
#!/usr/bin/env python3

import base64
import enum
import json

import kubernetes.client as k8s_client

from k8s import K8s
from .constants import GROUP
from .utils import camel_to_snake_case


class MonitorType(enum.Enum):
    HTTP = 1
    HTTPS = 1
    KEYWORD = 2
    PING = 3
    PORT = 4
    HEARTBEAT = 5


@enum.unique
class MonitorSubType(enum.Enum):
    HTTP = 1
    HTTPS = 2
    FTP = 3
    SMTP = 4
    POP3 = 5
    IMAP = 6
    CUSTOM = 99


@enum.unique
class MonitorHttpMethod(enum.Enum):
    HEAD = 1
    GET = 2
    POST = 3
    PUT = 4
    PATCH = 5
    DELETE = 6
    OPTIONS = 7


@enum.unique
class MonitorKeywordType(enum.Enum):
    EXISTS = 1
    NOT_EXISTS = 2


@enum.unique
class MonitorStatus(enum.Enum):
    PAUSED = 0
    NOT_CHECKED_YET = 1


@enum.unique
class MonitorHttpAuthType(enum.Enum):
    BASIC_AUTH = 1
    DIGEST = 2


@enum.unique
class MonitorPostType(enum.Enum):
    KEY_VALUE = 1
    RAW = 2


@enum.unique
class MonitorPostContentType(enum.Enum):
    TEXT_HTML = 0
    APPLICATION_JSON = 1


class MonitorV1Beta1:
    plural = 'uptimerobotmonitors'
    singular = 'uptimerobotmonitor'
    kind = 'UptimeRobotMonitor'
    short_names = ['urm']
    version = 'v1beta1'

    required_props = ['url', 'type']

    spec_properties = {
        'url': k8s_client.V1JSONSchemaProps(
            type='string',
            description='URL that will be monitored'
        ),
        'type': k8s_client.V1JSONSchemaProps(
            type='string',
            enum=list(
                MonitorType.__members__.keys()),
            description=f'Type of monitor, one of: {",".join(list(MonitorType.__members__.keys()))}'
        ),
        'friendlyName': k8s_client.V1JSONSchemaProps(
            type='string',
            description='Friendly name of monitor, defaults to name of UptimeRobotMonitor object'
        ),
        'subType': k8s_client.V1JSONSchemaProps(
            type='string',
            enum=list(
                MonitorSubType.__members__.keys()),
            description=f'Subtype of monitor, one of: {",".join(list(MonitorType.__members__.keys()))}'
        ),
        'port': k8s_client.V1JSONSchemaProps(
            type='integer',
            description=f'Port to monitor when using monitor sub type {MonitorType.PORT.name}'
        ),
        'keywordType': k8s_client.V1JSONSchemaProps(
            type='string',
            enum=list(
                MonitorKeywordType.__members__.keys()),
            description=f'Keyword type when using monitor type {MonitorType.KEYWORD.name}, one of: {",".join(list(MonitorKeywordType.__members__.keys()))}'
        ),
        'keywordValue': k8s_client.V1JSONSchemaProps(
            type='string',
            description=f'Keyword value when using monitor type {MonitorType.KEYWORD.name}'
        ),
        'interval': k8s_client.V1JSONSchemaProps(
            type='integer',
            multiple_of=60.,
            description='The interval for the monitoring check (300 seconds by default)'
        ),
        'httpUsername': k8s_client.V1JSONSchemaProps(
            type='string',
            description=f'Used for password protected pages when using monitor type {MonitorType.HTTP.name},{MonitorType.HTTPS.name} or {MonitorType.KEYWORD.name}, deprecated: use httpAuthSecret'
        ),
        'httpPassword': k8s_client.V1JSONSchemaProps(
            type='string',
            description=f'Used for password protected pages when using monitor type {MonitorType.HTTP.name},{MonitorType.HTTPS.name} or {MonitorType.KEYWORD.name}, deprecated: use httpAuthSecret'
        ),
        'httpAuthSecret': k8s_client.V1JSONSchemaProps(
            type='string',
            description=f'reference to a Kubernetes secret in the same namespace containing user and password for password protected pages when using monitor type {MonitorType.HTTP.name},{MonitorType.HTTPS.name} or {MonitorType.KEYWORD.name}'
        ),
        'httpAuthType': k8s_client.V1JSONSchemaProps(
            type='string',
            enum=list(
                MonitorHttpAuthType.__members__.keys()),
            description=f'Used for password protected pages when using monitor type {MonitorType.HTTP.name},{MonitorType.HTTPS.name} or {MonitorType.KEYWORD.name}, one of: {",".join(list(MonitorHttpAuthType.__members__.keys()))}'
        ),
        'httpMethod': k8s_client.V1JSONSchemaProps(
            type='string',
            enum=list(
                MonitorHttpMethod.__members__.keys()),
            description=f'The HTTP method to be used, one of: {",".join(list(MonitorHttpMethod.__members__.keys()))}'
        ),
        'postType': k8s_client.V1JSONSchemaProps(
            type='string',
            enum=list(
                MonitorPostType.__members__.keys()),
            description='The format of data to be sent with POST, PUT, PATCH, DELETE, OPTIONS requests'
        ),
        'postContentType': k8s_client.V1JSONSchemaProps(
            type='string',
            enum=list(
                MonitorPostContentType.__members__.keys()),
            description=f'The Content-Type header to be sent with POST, PUT, PATCH, DELETE, OPTIONS requests, one of: {",".join(list(MonitorPostContentType.__members__.keys()))}'
        ),
        'postValue': k8s_client.V1JSONSchemaProps(
            type='object',
            description='The data to be sent with POST, PUT, PATCH, DELETE, OPTIONS requests',
            x_kubernetes_preserve_unknown_fields=True
        ),
        'customHttpHeaders': k8s_client.V1JSONSchemaProps(
            type='object',
            description='Custom HTTP headers to be sent along monitor request, formatted as JSON',
            x_kubernetes_preserve_unknown_fields=True
        ),
        'customHttpStatuses': k8s_client.V1JSONSchemaProps(
            type='string',
            description='Allows to define HTTP status codes that will be handled as up or down, e.g. 404:0_200:1 to accept 404 as down and 200 as up'
        ),
        'ignoreSslErrors': k8s_client.V1JSONSchemaProps(
            type='boolean',
            description='Flag to ignore SSL certificate related issues'
        ),
        'alertContacts': k8s_client.V1JSONSchemaProps(
            type='string',
            description='Alert contacts to be notified when monitor goes up or down. For syntax check https://uptimerobot.com/api/#newMonitorWrap'
        ),
        'mwindows': k8s_client.V1JSONSchemaProps(
            type='string',
            description='Maintenance window IDs for this monitor'
        )
    }

    crd = k8s_client.V1CustomResourceDefinition(
        api_version='apiextensions.k8s.io/v1',
        kind='CustomResourceDefinition',
        metadata=k8s_client.V1ObjectMeta(name=f'{plural}.{GROUP}'),
        spec=k8s_client.V1CustomResourceDefinitionSpec(
            group=GROUP,
            versions=[k8s_client.V1CustomResourceDefinitionVersion(
                name=version,
                served=True,
                storage=True,
                schema=k8s_client.V1CustomResourceValidation(
                    open_apiv3_schema=k8s_client.V1JSONSchemaProps(
                        type='object',
                        properties={
                            'spec': k8s_client.V1JSONSchemaProps(
                                type='object',
                                required=required_props,
                                properties=spec_properties
                            ),
                            'status': k8s_client.V1JSONSchemaProps(
                                type='object',
                                x_kubernetes_preserve_unknown_fields=True
                            )
                        }
                    )
                )
            )],
            scope='Namespaced',
            names=k8s_client.V1CustomResourceDefinitionNames(
                plural=plural,
                singular=singular,
                kind=kind,
                short_names=short_names
            )
        )
    )

    @staticmethod
    def spec_to_request_dict(namespace:str, name: str, spec: dict) -> dict:
        k8s = K8s()

        # convert all keys from camel to snake case
        request_dict = {camel_to_snake_case(k): v for k, v in spec.items()}
        request_dict['friendly_name'] = request_dict.get('friendly_name', name)
        request_dict['type'] = MonitorType[spec['type']].value

        if 'http_auth_secret' in request_dict:
            secret = k8s.get_secret(namespace, request_dict['http_auth_secret'])

            request_dict['http_username'] = base64.b64decode(secret.data['username']).decode()
            request_dict['http_password'] = base64.b64decode(secret.data['password']).decode()
            request_dict.pop('http_auth_secret')

        # map enum values
        for key, enum_class in {
            'sub_type': MonitorSubType,
            'keyword_type': MonitorKeywordType,
            'http_auth_type': MonitorHttpAuthType,
            'http_method': MonitorHttpMethod,
            'post_type': MonitorPostType,
            'post_content_type': MonitorPostContentType
        }.items():
            request_dict[key] = enum_class[request_dict[key]
                                           ].value if key in request_dict else None

        # drop None entries

        return {k: v for k, v in request_dict.items() if v is not None}

    @staticmethod
    def annotations_to_spec_dict(annotations: dict) -> dict:
        spec = {}

        for key, value in annotations.items():
            if key not in MonitorV1Beta1.spec_properties:
                continue

            if MonitorV1Beta1.spec_properties[key].type == 'integer':
                spec[key] = int(value)
            elif MonitorV1Beta1.spec_properties[key].type == 'object':
                spec[key] = json.loads(value)
            else:
                spec[key] = value

        return spec

    @staticmethod
    def construct_k8s_ur_monitor_body(namespace, name=None, **spec):
        metadata = {
            'namespace': namespace
        }

        if name:
            metadata['name'] = name

        return {
            'apiVersion': f'{GROUP}/{MonitorV1Beta1.version}',
            'kind': MonitorV1Beta1.kind,
            'metadata': metadata,
            'spec': spec
        }


